dvc/ignore.py

```diff
@@ -0,0 +1,81 @@
+from fnmatch import fnmatch
+import os
+
+
+class DvcIgnore(object):
+    def __call__(self, *args, **kwargs):
+        assert len(args) == 3
+        root, dirs, files = args
+
+        return self._filter(root, dirs, files)
+
+    def _filter(self, root, dirs, files):
+        raise NotImplementedError
+
+
+class DvcIgnoreFromFile(DvcIgnore):
+    def __init__(self, ignore_file_path):
+        self.dirname = os.path.normpath(os.path.dirname(ignore_file_path))
+        self.patterns = self._read_lines_ignore_comments(ignore_file_path)
+
+    def _filter(self, root, dirs, files):
+
+        files = [f for f in files if not self.matches(root, f)]
+        dirs = [d for d in dirs if not self.matches(root, d)]
+
+        return dirs, files
+
+    def matches(self, root, basename):
+        full_path = os.path.join(root, basename)
+        rel_path = os.path.relpath(full_path, self.dirname)
+
+        for pattern in self.patterns:
+            if fnmatch(full_path, pattern) or fnmatch(rel_path, pattern):
+                return True
+
+        return False
+
+    def _read_lines_ignore_comments(self, file_path):
+        with open(file_path, "r") as fobj:
+            return [l for l in fobj if not l.lstrip().startswith("#")]
```

Comment:
If we don't support `!` syntax, it is a good idea to ignore such entries and/or put a warning and/or even error about it.